lib/doorkeeper/request.rb

```diff
@@ -3,6 +3,16 @@
 module Doorkeeper
   module Request
     class << self
+      def client_authentication_mechanism(request)
+        client_authentication_mechanisms = client_authentication_mechanisms.detect do |mechanism|
+          mechanism.matches_request?(request)
+        end
+
+        if client_authentication_mechanisms
+          client_authentication_mechanisms.mechanism
+        end
```

Comment:
I'm not sure what the "fallback" would be here? I think it'd just be a flow that returned empty `Credentials`?